### PR TITLE
ENH: Add SynthStrip base module

### PR DESCRIPTION
**Pull request by oesteban**
Translated the SynthStrip codebase into something more MRIQC-y or nipype-y, removing the dependency on the freesurfer bundle.

I have tried this on one image locally, stripping out the model from the freesurfer nightly build and it did work nicely.

Currently writing a nipype interface.

cc/ @ahoopes for feedback and comments/concerns

cc/ @satra

### CODE REVIEW

**Comment by oesteban** [comment]
BTW, Andrew, before merging I'll be more careful and make sure I mention and distribute the current license of SyntStrip and also mention my modifications as part of the apache licensing terms of MRIQC.

**Comment by oesteban** [comment]
#275, #554, #904

**Comment by oesteban** [comment]
Okay, this comes at a price I'm willing to pay:

RSS:
resources jqyrf1ei rss

VSM:
resources jqyrf1ei vsm

**Comment by ahoopes** [comment]
This all looks good to me. Regarding memory usage - I think pytorch might have the habit of using as much memory as it can. In theory, I think the model could run inference with much less if it needs to. Don't know for sure though, something to look into. My knowledge is mostly based on how tensorflow manages tensor allocation.

Anyway, we'll probably release a much smaller model at some point since we seem to retain accuracy pretty well after reducing network capacity. That might cut down on memory as well.

**Comment by oesteban** [comment]
@mgxd @eilidhmacnicol please have a quick look. I have tested on two pretty bad sites of ABIDE and it worked well in all the images (~100 cases I have checked visually).

File: mriqc/interfaces/synthstrip.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    model = File(
+        str(_default_model_path),
+        usedefault=True,
+        exists=True,
+        argstr="--model %s",
+        desc="file containing model's weights",
+    )
+    border_mm = traits.Int(
+        1, usedefault=True, argstr="-b %d", desc="Mask border threshold in mm"
+    )
```
Comment (eilidhmacnicol) [inline mriqc/interfaces/synthstrip.py:61]: Will the underlying command line call accept an argument of 0 here? (i.e. would you be able to set no threshold?)

File: mriqc/workflows/anatomical.py [hunk 2]
```diff
@@ -845,6 +834,102 @@ def airmsk_wf(name="AirMaskWorkflow"):
...
+
+    pre_n4 = pe.Node(
+        N4BiasFieldCorrection(
+            dimension=3,
+            num_threads=omp_nthreads,
+            rescale_intensities=True,
+            copy_header=True,
+        ),
+        name="pre_n4",
+    )
```
Comment (eilidhmacnicol) [inline mriqc/workflows/anatomical.py:862]: I know we are only certain this will work with humans, but thinking forward, would it be possible to make the n4 more configurable by adding function parameters? i.e. maybe setting bspline grids?

**Review by eilidhmacnicol — APPROVED** [review]
LGTM - just a few points but only in regards to future options for expansion to other species (so take them with a pinch of salt).

It's possible to merge these now without perturbing the rodent workflow (so if that's the only reason why you requested my review then go for it) but if we wanted to test it on rodents in future, the synthstrip workflow might need some more flexibility for n4 parameters.

File: mriqc/interfaces/synthstrip.py (hunk 1 above)
Comment (oesteban) [reply mriqc/interfaces/synthstrip.py:61]: Yes, that would be trivial.

File: mriqc/workflows/anatomical.py (hunk 2 above)
Comment (oesteban) [reply mriqc/workflows/anatomical.py:862]: Sure, but first we will need to make sure SynthStrip works on rodents...

**Comment by oesteban** [comment]
Okay, this is going in - I'm not looking back.

File: mriqc/workflows/anatomical.py (hunk 2 above)
Comment (ahoopes) [reply mriqc/workflows/anatomical.py:862]: Right, not sure how that would generalize. Might do OK, but not perfect. A multi-species training set would be a good next step though
